gpu/segmentation/include/pcl/gpu/segmentation/impl/gpu_extract_clusters.hpp

```diff
@@ -132,35 +162,34 @@ pcl::gpu::extractEuclideanClusters (const pcl::PointCloud<pcl::PointXYZ>::Ptr  &
       // If number of queries is high enough do it here
       else
       {
-        std::cout << " GPU: ";
+        PCL_DEBUG(" GPU: ");
         // Copy buffer
         queries_device = DeviceArray<PointXYZ>(queries_device_buffer.ptr(),queries_host.size());
         // Move queries to GPU
         queries_device.upload(queries_host);
         // Execute search
         tree->radiusSearch(queries_device, tolerance, max_answers, result_device);
         // Copy results from GPU to Host
-        result_device.sizes.download (sizes);
-        result_device.data.download (data);
+        result_device.sizes.download(sizes);
+        gpu_indices.clear();
+        pcl::detail::economical_download(result_device, sizes, max_answers, gpu_indices);
         // Store the previously found number of points
         previous_found_points = found_points;
         // Clear queries list
         queries_host.clear();
-        for(std::size_t qp = 0; qp < sizes.size (); qp++)
-        {
-          for(int qp_r = 0; qp_r < sizes[qp]; qp_r++)
-          {
-            if(processed[data[qp_r + qp * max_answers]])
-              continue;
-            processed[data[qp_r + qp * max_answers]] = true;
-            queries_host.push_back ((*host_cloud_)[data[qp_r + qp * max_answers]]);
+        for (int idx : gpu_indices) {
+            if (processed[idx])
+                continue;
+            processed[idx] = true;
+            queries_host.push_back((*host_cloud_)[idx]);
             found_points++;
-            r.indices.push_back(data[qp_r + qp * max_answers]);
-          }
+            r.indices.push_back(idx);
         }
```

Comment:
2 thoughts:
* this is exactly the same as L120-130, thanks to economical download
* L189 will give wrong result since data is not used in this branch of if-else, instead gpu_indices are used

Reply:
Ha! You pointed out an intriguing symmetry (if I understood what you meant). We can indeed refactor the code nicely. I made an attempt to do that. What do you think about it?